### DISCUSSION TITLE
[Share] Insert-Template 更新Note中期刊分区/IF等信息

### DISCUSSION THREAD

**Discussion by ramonmi** [windingwind/zotero-actions-tags#341] [Action Scripts]
### Description

每年年中科睿唯安会发布期刊的JCR分区和影响因子（IF），每年年底中科院会发布期刊的中科院分区（升级版）等，这导致我们使用Better Note模板生成的笔记中的期刊分区/IF信息必然存在过期的时候，对此提出了一个解决方案。
1. 每次科睿唯安/中科院更新期刊分区/IF后，使用Green Frog或（Zotero Style，可能需要修改Template）插件更新期刊item的相关信息
2. 将Better Note Template中摘取期刊信息的部分单独作为一个[Item]模板，例如 https://github.com/windingwind/zotero-better-notes/discussions/1016 中 Anything else 的 [Item] update_IF_CAS 更新模板
3. 运行脚本需要选中想要更新期刊信息的Note，在保持选中Note item的前提下，可在以Tab窗口打开（open as tab）的窗口运行该脚本，但不能在open as window或者在单独窗口打开（Edit in a Separate Window）时运行，否则脚本无法生效

尽管打开Note后，点击右上角三个点，选择insert template也可以实现相应功能，但是不仅需要选择note template，还需要选择regular item才可以插入，步骤繁琐

If you want to reset the remembered template, go to preferences - advanced - config editor - search 
 `extensions.actionsTags.customAction.insertBNTemplate` and reset.

在第一次运行动作时弹窗选择需要的笔记模板。如果已经运行过一次，要换成其他笔记模板，则需要通过高级设置修改 `extensions.actionsTags.customAction.insertBNTemplate` 为目标笔记模板的名称。

Require Better Notes for Zotero >= 1.1.4-21

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Insert note to the current note using the default Better Notes template 
 * @author windingwind, ramonmi
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/341
 * @refs https://github.com/windingwind/zotero-actions-tags/discussions/108
 * @usage Trigger with shortcuts
 * @Hint must set the line number according to the actual situation
 */

const Zotero = require("Zotero");
const window = require("window");

if (!Zotero.BetterNotes?.api.note?.insert) {
	return "[Action:Insert-Template] Better Notes for Zotero > 1.1.4-21 is not installed or disabled.";
}
if (!item) {
	return "[Action:Insert-Template] Target item is empty";
}

const prefsKey = "extensions.actionsTags.customAction.insertBNTemplate";

// Get template name
let templateName = Zotero.Prefs.get(prefsKey, true);
let templateContent = Zotero.BetterNotes.api.template.getTemplateText(templateName);
if (!templateContent) {
	// Template name is not defined or invalid. Get from input
	const templateNames = Zotero.BetterNotes.api.template.getTemplateKeys();

	function selectItems(itemList) {
		var io = { dataIn: itemList, dataOut: null };
		window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
			"_blank", "chrome,modal,centerscreen,resizable=yes", io);
		return io.dataOut;
	}

	templateName = Object.values(selectItems(templateNames))[0];
	templateContent = Zotero.BetterNotes.api.template.getTemplateText(templateName);
	if (!templateContent) {
		return "[Action:Insert-Template] Template is invalid";
	}
	Zotero.Prefs.set(prefsKey, templateName, true);
}

const parentItem = Zotero.Items.getTopLevel([item])[0];

if (parentItem.isRegularItem()&&item.isNote()) {
	let html = "";
	if (templateName.toLowerCase().startsWith("[item]")) {
		html = await Zotero.BetterNotes.api.template.runItemTemplate(templateName, {
			itemIds: [parentItem.id],
			targetNoteId: item.id,
		});
	} else {
		html = await Zotero.BetterNotes.api.template.runTextTemplate(templateName, {
			targetNoteId: item.id,
		});
	}
	await Zotero.BetterNotes.api.note.insert(
		item,
		html,
		5, // Must set the line number according to your Note or template
	);
	return `[Action:Insert-Template] Note is inserted to the note ${item.getNoteTitle()} from template ${templateName}`;
} else {
	return `[Action:Insert-Template] Please check if the selected item is a Childnote item!`
}
```

### Anything else

- 以上解决方案仅提供一个思路，需根据个人的Note模板进行修改，特别是需要更改note插入行的行号（默认第5行，基于 https://github.com/windingwind/zotero-better-notes/discussions/1016 ）
- 仍有较大提升空间，例如是否可以实现对应行替换，抑或是插入到光标所在位置，但似乎只有editor中有`getLineAtCursor`，但未找到对应api，还希望 @windingwind 可以指点一二，也希望大家可以提出更好的解决方案

**Comment by kdqiu** [comment]
请问如何重新选择新的note模板，我运行了一次之后就无法选择新的模板

**Reply by kdqiu** [reply]
已找到解决办法，感谢。来自[108](https://github.com/windingwind/zotero-actions-tags/discussions/108)

**Reply by ramonmi** [reply]
嗯嗯，文中也摘录了 @windingwind 的修改模板的方法介绍，只不过是英文的，现已添加上中文。
Better Notes最新的release https://github.com/windingwind/zotero-better-notes/releases/tag/v1.1.4-beta.100 添加了 update content from template (beta) 功能，期待功能得到完善，这样似乎就可以直接在Note界面选择模板进行更新了

**Reply by llity** [reply]
> 嗯嗯，文中也摘录了 @windingwind 的修改模板的方法介绍，只不过是英文的，现已添加上中文。 Better Notes最新的release https://github.com/windingwind/zotero-better-notes/releases/tag/v1.1.4-beta.100 添加了 update content from template (beta) 功能，期待功能得到完善，这样似乎就可以直接在Note界面选择模板进行更新了

不兼容夜间模式。

**Reply by ramonmi** [reply]
是actions-tag不兼容还是Better-notes不兼容呢？建议在对应插件新开issue反馈这个问题

**Reply by ramonmi** [reply]
> 嗯嗯，文中也摘录了 @windingwind 的修改模板的方法介绍，只不过是英文的，现已添加上中文。 Better Notes最新的release https://github.com/windingwind/zotero-better-notes/releases/tag/v1.1.4-beta.100 添加了 update content from template (beta) 功能，期待功能得到完善，这样似乎就可以直接在Note界面选择模板进行更新了

- 利用Better Notes 插件的 `update content from template` 功能，已实现**更好的每年更新期刊分区/IF信息的解决方法**，可参见以下链接
    - New Template for Better Notes: https://github.com/windingwind/zotero-better-notes/discussions/1256
    - 利用BN插件API，将旧Notes中的期刊信息部分替换为新的**可刷新期刊数据**：https://github.com/windingwind/zotero-actions-tags/discussions/448
- 如有更好的方案，欢迎大家讨论！感谢 @windingwind 的指导！
